examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+      values(c) = BoundaryValues<dim>::value (p, c);
+  }
+
+
+  // @sect3{BlockSchurPreconditioner for Navier Stokes equations}
+  //
+  // The block
+  // Schur complement preconditioner is defined in this part. As discussed in
+  // the introduction, the preconditioner in Krylov iterative methods is
+  // implemented as a matrix-vector product operator. In practice, the Schur
+  // complement preconditioner is decomposed as a product of three matrices(as
+  // presented in the first section).  The $\tilde{A}^{-1}$ in the first factor
+  // involves a solve for the linear system $\tilde{A}x=b$. Here we solve
+  // this system via a direct solver for simplicity. The computation involved
+  // in the second factor is a simple matrix-vector multiplication. The Schur
+  // complement $\tilde{S}$ can be well approximated by the pressure mass
+  // matrix and its inverse can be obtained through an inexact solver. Because
+  // the pressure mass matrix is symmetric and positive definite, we can use
+  // CG to solve the corresponding linear system.
+  //
+  template <class PreconditionerMp>
+  class BlockSchurPreconditioner : public Subscriptor
+  {
+  public:
+    BlockSchurPreconditioner (double                                     gamma,
+                              double                                     viscosity,
+                              const BlockSparseMatrix<double>            &S,
+                              const SparseMatrix<double>                 &P,
+                              const PreconditionerMp                     &Mppreconditioner
+                             );
+
+    void vmult (BlockVector<double>       &dst,
+                const BlockVector<double> &src) const;
+
+  private:
+    const double gamma;
+    const double viscosity;
+    const BlockSparseMatrix<double> &stokes_matrix;
+    const SparseMatrix<double> &pressure_mass_matrix;
+    const PreconditionerMp &mp_preconditioner;
+    SparseDirectUMFPACK A_direct;
+  };
+
+  template <class PreconditionerMp>
+  BlockSchurPreconditioner<PreconditionerMp>::
+  BlockSchurPreconditioner (double                           gamma,
+                            double                           viscosity,
+                            const BlockSparseMatrix<double>  &S,
+                            const SparseMatrix<double>       &P,
+                            const PreconditionerMp           &Mppreconditioner)
+    :
+    gamma                (gamma),
+    viscosity            (viscosity),
+    stokes_matrix        (S),
+    pressure_mass_matrix (P),
+    mp_preconditioner    (Mppreconditioner)
+  {
+    A_direct.initialize(stokes_matrix.block(0,0));
+  }
+
+  template <class PreconditionerMp>
+  void
+  BlockSchurPreconditioner<PreconditionerMp>::
+  vmult (BlockVector<double>       &dst,
+         const BlockVector<double> &src) const
+  {
+    Vector<double> utmp(src.block(0));
+
+    {
+      SolverControl solver_control(1000, 1e-6 * src.block(1).l2_norm());
```

Comment:
Does this match the stated tolerance in the introduction? I thought that there you say 1e-4.